Allow removing TextTracks created by hls.js

xxoo opened the pull request:
Allow removing TextTracks created by hls.js
### This PR will...

- use `appendChild` instead of `addTextTrack` to allow reverse operation
- replace track flush and reuse workflow with simple removal
- add `trackNode` field in `MediaPlaylist` to indicate associated `TextTrack`
- fix `MEDIA_DETACHED` event in wrong order and sometimes not triggering

### Why is this Pull Request needed?

This PR allows video elements to have the same lifecycle as hls.js instances. After detaching, you may attach them to new hls.js instances with nothing to worry about. It also fixed a bug caused by track reuse, where content overlapping occurs if a video contains multiple captions sharing the same language and label. eg: https://devstreaming-cdn.apple.com/videos/streaming/examples/bipbop_16x9/bipbop_16x9_variant.m3u8

### Are there any points in the code the reviewer needs to double check?

Yes. Since the changes involve more than one place, extensive testing from multiple aspects may be necessary to ensure no issues are left behind. Especially I'm not sure whether it works well with `transferMedia()`

### Resolves issues:
https://github.com/video-dev/hls.js/issues/2198

### Checklist

- [x] changes have been done against master branch, and PR does not conflict
- [x] new unit / functional tests have been added (whenever applicable)
- [x] API or design changes are documented in API.md

robwalch commented on src/utils/texttrack-utils.ts:
```diff
@@ -1,16 +1,45 @@
   }
-  (event as any).track = track;
-  videoEl.dispatchEvent(event);
+  // To avoid an issue of the built-in captions menu in Chrome
+  if (navigator.userAgent.includes('Chrome/')) {
+    el.src = 'data:,WEBVTT';
+  }
```
Interesting.

Is this to prevent loading of the current page (src="" <identity> treated like "." path)? 
What other issues occur with the menu?

I recall there was some issue in Safari with DOM created tracks but it has been long enough that I don't remember the details.

xxoo replied:
So far, I haven't seen any issue related to DOM tracks on Safari. However, the [issue](https://github.com/video-dev/hls.js/issues/2198#issuecomment-761614248) you found on Chrome still exists. Providing a valid VTT file might help to avoid this, and the data URL approach serves exactly this purpose.

Subtitle tracks not managed by hls.js shouldn't cause problems, or at least such problems can be avoided. Because hls.js keeps a reference to each track it manages. If a user selects an unmanaged track, hls.js will leave it as is and set the current subtitle track index to -1. I think I can add some unit tests for this scenario.

According to caniuse.com, the [track](https://caniuse.com/?search=track) element and the [addTextTrack](https://caniuse.com/?search=addTextTrack) method have nearly identical browser support. The only difference I've noticed is that a newly created track node might need some time to initialize its mode. I'm not sure if this behavior affects any public API usage. Perhaps we should further analyze the necessity and implications of this, as maintaining the old approach could complicate the workflow.

robwalch replied:
> The only difference I've noticed is that a newly created track node might need some time to initialize its mode.

We should also have a look at whether or not these changes impact performance WRT #7093. I'd be curious to know if adding tracks this way has the same cost in Safari or not. If it resolved the issue without any changes to WebKit that would be another +1 for accepting the change.

robwalch replied:
> performance WRT https://github.com/video-dev/hls.js/issues/7093

No difference. There appears to be a fixed cost for each track added in Safari that can be seen in the profiling results under "Add Track Event Dispatched" Timelines tab of the Web Inspector.

robwalch reviewed (changes requested):
For backwards compatibility we need to maintain `addTextTrack` as the default path for adding TextTracks.

There may be JavaScript platforms with limited Web support cannot create TextTrack nodes. There are also browsers where media.textTracks from DOM created tracks do not behave the same (HTMLMediaElement controls menu issues, events, etc...). One concern is that apps may provide an element with tracks already added by another means or they may add their own side car tracks and this change would introduce problems with those integrations.

Would be possible to make this change optional via a configuration option (use `addTextTrack` by default, but set an option to use DOM tracks instead)?

robwalch commented on src/controller/subtitle-track-controller.ts:
```diff
@@ -149,16 +146,15 @@ class SubtitleTrackController extends BasePlaylistController {
-    textTracks.forEach((track) => {
-      clearCurrentCues(track);
-    });
+    for (const track of this.tracks) {
```
Please use for loops or forEach method, matching the syntax in existing code.

"for of" loops get replaced by babel with iterator functions (`_createForOfIteratorHelperLoose`) that add additional boilerplate to the ES5-UMD library output.

If you'd rather not, or see them in a couple of other places, we can follow up with a task to add a lint rule (we shouldn't have unspoken rules) https://www.npmjs.com/package/eslint-plugin-no-for-of-loops

satyam73 replied:
Hi @robwalch 
Can I add this rule by creating separate PR?
Actually I wanted to start contributing to this repo and I think it will be a good small start for me

robwalch replied:
👋 @satyam73
Yes please. Go for it 🙌

satyam73 replied:
Sure I'll do it 👍
Will raise a PR in 1-2 days

robwalch commented on src/controller/subtitle-track-controller.ts:
```diff
@@ -170,12 +166,33 @@ class SubtitleTrackController extends BasePlaylistController {
 
+  private createTextTracks() {
+    const tracks = this.tracks;
+    for (const track of tracks) {
```
Please use for loops or forEach method, matching the syntax in existing code.

robwalch commented on src/controller/timeline-controller.ts:
```diff
@@ -280,43 +249,43 @@ export class TimelineController implements ComponentAPI {
+    const media = data.media;
+    this.media = media;
+    if (this.config.renderTextTracksNatively) {
+      for (const track of this.tracks) {
```
Please use for loops or forEach method, matching the syntax in existing code.

robwalch commented on src/controller/timeline-controller.ts:
```diff
@@ -280,43 +249,43 @@ export class TimelineController implements ComponentAPI {
-      delete captionsTracks[trackName];
-    });
+    if (media && this.config.renderTextTracksNatively) {
+      for (const track of this.tracks) {
```
Please use for loops or forEach method, matching the syntax in existing code.

robwalch commented on src/controller/timeline-controller.ts:
```diff
@@ -280,43 +249,43 @@ export class TimelineController implements ComponentAPI {
+        media.removeChild(track.trackNode!);
+      }
+      const { captionsTracks } = this;
+      for (const trackName in captionsTracks) {
```
Please use for loops or forEach method, matching the syntax in existing code.

robwalch commented on src/controller/timeline-controller.ts:
```diff
@@ -368,67 +320,24 @@ export class TimelineController implements ComponentAPI {
-            if (inUseTrack) {
-              textTrack = inUseTrack;
+        if (media) {
+          for (const track of this.tracks) {
```
Please use for loops or forEach method, matching the syntax in existing code.

robwalch commented on src/controller/timeline-controller.ts:
```diff
@@ -368,67 +320,24 @@ export class TimelineController implements ComponentAPI {
-            );
-            if (textTrack) {
-              textTrack.mode = 'disabled';
+          for (const track of tracks) {
```
Please use for loops or forEach method, matching the syntax in existing code.

robwalch commented on src/controller/timeline-controller.ts:
```diff
@@ -711,20 +621,22 @@ export class TimelineController implements ComponentAPI {
-            endOffsetSubtitles,
-          ),
-        );
+        for (const track of this.tracks) {
```
Please use for loops or forEach method, matching the syntax in existing code.

robwalch commented:
> According to caniuse.com, the [track](https://caniuse.com/?search=track) element and the [addTextTrack](https://caniuse.com/?search=addTextTrack) method have nearly identical browser support.

That does not guarantee that these changes will not break someone's integration. There could be integrations of hls.js out there that depend at some level on `media.addTextTrack` getting called. We can't know until releasing these changes, but don't be surprised. Someone might come out of nowhere saying that `ownerDocument` in undefined on their media element.

That said, I marked this under milestone v1.7. This is really nice work.

xxoo commented:
Thanks for the review. I will fix these issues soon. I also think removable tracks may help improve performance slightly when there are too many text tracks and more than one text group, as we can destroy tracks outside current groups. Shaka Player avoids this problem by adding only one text track, meaning the native captions UI or API isn't available as an option for users.

xxoo commented:
@robwalch 
I just updated this PR with the following changes:
- Replace for...of with forEach
- Only create track nodes for current groups
- No longer depend on event order
- Adjusted the way hls.js handles external subtitles to fix unexpected behavior

You may reproduce the issue on master branch with the unit test "should set subtitleTrack to -1 and keep unmanagedTrack showing"

robwalch commented on src/controller/subtitle-track-controller.ts:
```diff
@@ -546,10 +543,8 @@ class SubtitleTrackController extends BasePlaylistController {
     }
     if (!track) {
-      if (lastTrack) {
-        // switch to -1
```
`SUBTITLE_TRACK_SWITCH` is only expected to trigger when there is a track change. The thinking here is that if `this.currentTrack` (`lastTrack`) is `null` then it is not set , `trackId` must also be -1 already, and setting `subtitleTrack` to -1 is a no-op.

xxoo replied:
Yes this detection was add by me. But it breaks the unit test "should trigger SUBTITLE_TRACK_SWITCH if passed -1" (which I removed before for easier debugging, but forgot to add it back).
 In the past, setting to -1 when it is already -1 may still change the `mode` of external subtitles in `toggleTrackModes()`. But this behavior has changed now. Which makes setting to -1 in this situation have no effect. I change them back only for passing this unit test. Although I totally agree with you. If this unit test is no longer necessary (I believe it is true, but maybe it is not for me to decide), I'd also like to add this detection back.

robwalch commented on tests/unit/controller/timeline-controller.js:
```diff
@@ -5,177 +5,25 @@ import { Events } from '../../../src/events';
-    });
-  });
-
-  describe('text track kind', function () {
```
Why were the track `kind` tests removed?

xxoo replied:
Because `TimelineController` no longer creates `TextTrack` for `MediaPlaylist`. Instead, the life cycle of `<track>` elements are managed by `SubtitleTrackController` now. There is a equivalent test in [tests/unit/controller/subtitle-track-controller.ts](https://github.com/video-dev/hls.js/pull/7184/files/8754b2efff5af43f607a68d9b813f1e6b7ac6aa1#diff-e9cc5c36fc5450e8a9bdf1b6cc6d21ee18d85caee04840acca56fd82b7c4d95bR187)

robwalch commented:
Hi @xxoo,

Sorry for the delay in getting to this. Can you please resolve the merge conflicts by rebasing these changes?

xxoo commented:
Sorry seems I used merge in some step. Maybe I'll have to start a new PR to fix this.
